clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreBufferService.java

```diff
@@ -472,138 +201,421 @@ public void drainBufferedRecordsFromTopicPartition(PubSubTopicPartition topicPar
-      StoreBufferDrainer drainer = drainerList.get(index);
-      // print drainer info when there is a slow drainer.
-      if (slowDrainerExists) {
-        MemoryBoundBlockingQueue<QueueNode> queue = blockingQueueArr.get(index);
-        int count = queue.getMemoryUsage() > 0.8 * bufferCapacityPerDrainer ? 5 : 1;
-        List<Map.Entry<PubSubTopicPartition, Long>> slowestEntries = drainer.topicToTimeSpent.entrySet()
-            .stream()
-            .sorted(comparing(Map.Entry::getValue, reverseOrder()))
-            .limit(count)
-            .collect(toList());
-
-        int finalIndex = index;
-        slowestEntries.forEach(
-            entry -> LOGGER
-                .info("In drainer number {}, time spent on {} : {} ms", finalIndex, entry.getKey(), entry.getValue()));
-      }
-      drainer.topicToTimeSpent.clear();
+    public void stop() {
+      isRunning.set(false);
     }
 
-    return maxUsage;
-  }
+    @Override
+    public void run() {
+      LOGGER.info("Starting StoreBufferDrainer Thread for drainer: {}....", drainerIndex);
+      QueueNode node = null;
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord = null;
+      LeaderProducedRecordContext leaderProducedRecordContext = null;
+      StoreIngestionTask ingestionTask = null;
+      CompletableFuture<Void> recordPersistedFuture = null;
+      while (isRunning.get()) {
+        try {
+          node = blockingQueue.take();
 
-  /** Used for testing */
-  Map<PubSubTopicPartition, Long> getTopicToTimeSpentMap(int i) {
-    return drainerList.get(i).topicToTimeSpent;
-  }
+          consumerRecord = node.getConsumerRecord();
+          leaderProducedRecordContext = node.getLeaderProducedRecordContext();
+          ingestionTask = node.getIngestionTask();
+          recordPersistedFuture = node.getQueuedRecordPersistedFuture();
 
-  @Override
-  public long getMinMemoryUsagePerDrainer() {
-    long minUsage = Long.MAX_VALUE;
-    for (MemoryBoundBlockingQueue<QueueNode> queue: blockingQueueArr) {
-      minUsage = Math.min(minUsage, queue.getMemoryUsage());
+          long startTime = System.currentTimeMillis();
+
+          int subPartition = PartitionUtils
+              .getSubPartition(consumerRecord.getTopicPartition(), ingestionTask.getAmplificationFactor());
+
+          processRecord(
+              consumerRecord,
+              ingestionTask,
+              leaderProducedRecordContext,
+              subPartition,
+              node.getKafkaUrl(),
+              node.getBeforeProcessingRecordTimestampNs());
+
+          /**
+           * Complete {@link QueueNode#queuedRecordPersistedFuture} since the processing for the current record is done.
+           */
+          if (recordPersistedFuture != null) {
+            recordPersistedFuture.complete(null);
+          }
+          long latencyInMS = System.currentTimeMillis() - startTime;
+          this.stats.recordInternalProcessingLatency(latencyInMS);
```

Comment:
This might be a dumb question but what do we use `internalProcessingLatency` for? The metric seems to cover a variety of behaviors since `processRecord` can be called on control messages and data records. The expected latency can be very different for different messages. Especially if message is rejected/skipped by DIV the latency will be really low. Basically not sure how this metric can be useful when any value could be either good or bad.

Reply:
Good question! It's true that this metric doesn't differentiate records and we probably cannot tell whether slowness happens in which behavior (in that case, I think we need to add a new metric for that behavior or check different stats, e.g. `min` for control message 
 or `max` for large data record). However, it gives us the insight on how fast a drainer can process the record. Right now, we don't have the data and existing end-to-end metrics include the drainer queueing time which makes them unreliable to understand the internal processing latency. With this metric, we can also deduce how long a record is queued in the drainer roughly..

Reply:
I see, but data record's process latency should mostly be storage engine put latency which we already have a metric for. Given that this is not a store level metric I guess it's fine to add it and see the behavior.

Reply:
yes, I agree that storage engine put would be the major contributor. As we potentially add more codes to the internal processing path, this new metric gives some insights on how other code paths perform.